### PR TITLE
Counts reset if required

### LINKED ISSUE #6587: Reviewer - Negative Review Count after Undo
###### Reproduction Steps

1.  Create new deck
2.  Add two-three cards
3.  Review card
4. Undo Review

###### Expected Result
It should reset review to initial state.

###### Actual Result
It reset but in minus.
desktop

###### Debug info
AnkiDroid 2.12alpha18

###### Research
*Enter an [x] character to confirm the points below:*

- [ x ] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid
- [ x ] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [ x ] I have searched for similar existing issues here and on the user forum
- [ x ] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
## Purpose / Description

This should solve
Fixes #6587 while still
doing resets only when they are required.

## Approach

The main problem is that "counts" mean "number of cards to see today, ignoring current card if any".

The trouble here is that "current card" was not something the scheduler new about. So instead we have ad-hoc code to ignore the current card. Essentially, when "getCard" is called, it is assumed it is giving the current card. When "undo" is called, it reset the scheduler twice and manually change the count of the current card.

The good news is that https://github.com/ankidroid/Anki-Android/pull/6585 is already dealing with saving current card. So this can be easily used to correct the counting. When counting is reset, we take current card into account to know what to change. 

Of course, we may just consider https://github.com/ankidroid/Anki-Android/pull/6585 first and then consider the resetting of counts later, to make this PR simpler.

## How Has This Been Tested?

I'm testing it on my phone, trying to reproduce the bug

## Learning (optional, can help others)

How to use bisect to find where a bug is introduced. And that tests which deal only with scheduler does not simulate the reviewer very corrrectly

### CODE REVIEW

**Comment by mikehardy** [comment]
Oh yes! Forgot those, but the more auto-testing the better, that's really important. I will wait to merge until those are pushed in

**Comment by Arthur-Milchior** [comment]
Done

File: AnkiDroid/src/test/java/com/ichi2/libanki/sched/AbstractSchedTest.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import java.util.Arrays;
+
+import static com.ichi2.anki.AbstractFlashcardViewer.EASE_3;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.not;
+
+// Note: These tests can't be run individually but can from the class-level
+// gradlew AnkiDroid:testDebug --tests "com.ichi2.libanki.sched.AbstractSchedTest.*"
+@RunWith(ParameterizedRobolectricTestRunner.class)
```
Comment (mikehardy) [inline AnkiDroid/src/test/java/com/ichi2/libanki/sched/AbstractSchedTest.java:41]: Love the multi-scheduler running @david-allison-1 !

**Comment by mikehardy** [comment]
Excellent thanks @Arthur-Milchior just waiting on CI now, once it's done processing it's current pile I'll ship 2.12alpha21 and we'll see how it goes - should be the branch point for real now I think

File: AnkiDroid/src/test/java/com/ichi2/libanki/sched/AbstractSchedTest.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        Card cardBeforeUndo = sched.getCard();
+        sched.setCurrentCard(cardBeforeUndo);
+        int[] countsBeforeUndo = sched.counts();
+        // Not shown in the UI, but there is a state where the card has been removed from the queue, but not answered
+        // where the counts are decremented.
+        assertThat(countsBeforeUndo, is(new int[] { 0, 0, 0 }));
+
+        sched.answerCard(cardBeforeUndo, EASE_3);
+
+        CollectionTask.launchCollectionTask(CollectionTask.TASK_TYPE_UNDO,
```
Comment (david-allison-1) [inline AnkiDroid/src/test/java/com/ichi2/libanki/sched/AbstractSchedTest.java:81]: Does this block until the task is completed?

File: AnkiDroid/src/test/java/com/ichi2/libanki/sched/AbstractSchedTest.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                        int[] countsAfterUndo = sched.counts();
+                        assertThat("Counts after an undo should be the same as before an undo", countsAfterUndo, is(countsBeforeUndo));
+
+                    }
+
+
+                    @Override
+                    public void onProgressUpdate(CollectionTask task, CollectionTask.TaskData... values) {
+                        Card undoneCard = (Card)values[0].getCard();
+                        sched.setCurrentCard(undoneCard);
```
Comment (david-allison-1) [inline AnkiDroid/src/test/java/com/ichi2/libanki/sched/AbstractSchedTest.java:100]: Why is this not done in the background in the task?

**Comment by Arthur-Milchior** [comment]
I just edited the main message to explain the reason of this change

File: AnkiDroid/src/test/java/com/ichi2/libanki/sched/AbstractSchedTest.java (hunk 3 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/test/java/com/ichi2/libanki/sched/AbstractSchedTest.java:100]: This is simulating the reviewer. The reviewer does this in onProgessUpdate.

File: AnkiDroid/src/test/java/com/ichi2/libanki/sched/AbstractSchedTest.java (hunk 2 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/test/java/com/ichi2/libanki/sched/AbstractSchedTest.java:81]: The actual test is done when the task is completed. Don't know whether it is your question

**Comment by Arthur-Milchior** [comment]
Since "counts" is "number of card to review, not counting the one currently in reviewer", I do assume that the easiest way is indeed to save the current card.

There is another solution, which would be even better I believe, ensuring that "counts" returns "number of cards to see according to current state of the collection" and letting the reviewer deal with the notion that "if current card is new and the scheduler states there are N new cards to review today, then the user will see N-1 cards after this one, so I'll display N-1". 

There should be no problem in doing it, except in the card content provider. If DECK_SELECTED is called while the reviewer is open, it would change the number of card it receive

**Comment by Arthur-Milchior** [comment]
On discord, @david-allison-1 stated:
>If the tests need to be from the perspective of the Reviewer - that's not an issue, but the Scheduler should ideally be the source of truth and state
>
>We have to take into account API clients that will exercise the Scheduler without the Reviewer
>But, it's hard to work around the constraints that libAnki provides

I do agree. So either we decide to stop following libanki here (or provide a method "counts-upstream" for the card content provider and a method "counts-independant of reviewer") for the remaining of the code (I'd gladly code it)
Or we accept that scheduler interact with reviewer in the complex ways I described above

**Comment by david-allison-1** [comment]
> On discord, @david-allison-1 stated:
> 
> > If the tests need to be from the perspective of the Reviewer - that's not an issue, but the Scheduler should ideally be the source of truth and state
> > We have to take into account API clients that will exercise the Scheduler without the Reviewer
> > But, it's hard to work around the constraints that libAnki provides
> 
> I do agree. So either we decide to stop following libanki here (or provide a method "counts-upstream" for the card content provider and a method "counts-independant of reviewer") for the remaining of the code (I'd gladly code it)
> Or we accept that scheduler interact with reviewer in the complex ways I described above

The problem is that `Sched` is no longer the source of state, and there appears to be a change in the externally visible state of `Sched` (`counts()`) unless the newly provided API call is used. 

Changing the tests to move away from testing `Sched`, and testing the via the externally visible behaviour/state of `Reviewer` can simplify things (those tests are far less brittle and will provide more long-term value), but we should avoid making a change in publicly visible behaviour in `Sched` as there are other clients which rely on it. This is even more true if the code change was intended to be a performance boost, rather than a planned functionality change.

As for what we do, I'll mull over the design. deferring the counts mutation seemed to me to be the right move if we want to be able to peek the queues and rebuild them early. If we can't do that without affecting the public API, we get a cache hit 49/50 of the time by prefetching the head elements on each of the queues.

**Comment by Arthur-Milchior** [comment]
I just found a really simple change that does not require to change the scheduler interface externally: "undo" sends the card to both the reviewer and the scheduler. This way, there is no need for reviewer to set current card anymore while "undoing"

Since this PR and https://github.com/ankidroid/Anki-Android/pull/6597 only change counting during resets, everything should work correctly here.

The reviewer still need to discard the card, because there is no way for the scheduler to know whether the card it sent is still considered by the user

**Comment by Arthur-Milchior** [comment]
I did a new version of the change you probably will prefer. It is very close to original code. Instead of saving the current card, `deferReset` simply saves the card that was undone, so the action of decrementing the count can occur after the reset occurs. If another reset occur, it means that this card is not in the reviewer anymore, so we just set the card to null.

**Pull request merged**